src/commands/fetch-balance.cpp

```diff
@@ -53,18 +61,36 @@ console_result fetch_balance::invoke(std::ostream& output, std::ostream& error)
 
     callback_state state(error, output, encoding);
 
-    auto on_done = [&state, &address](const code& ec,
+    auto on_done = [&state, &hash, &address](const code& ec,
         const history::list& rows)
     {
         if (!state.succeeded(ec))
             return;
 
         // This override summarizes the history response as balance.
-        state.output(prop_tree(rows, address));
+        if (hash == null_hash)
+            state.output(prop_tree(rows, address));
+        else
+            state.output(prop_tree(rows, hash));
     };
 
+    // Address is ignored if both are specified.
+    // Use the null_hash as sentinel to determine whether to use address or hash.
     // This does not include unconfirmed transactions.
-    client.blockchain_fetch_history4(on_done, address);
+    if (hash == null_hash)
+    {
+        client.blockchain_fetch_history4(on_done, address);
+    }
+    else
+    {
+        // If a hash option is provided directly, it's been reversed already and
+        // needs to be undone.
+        auto reversed_hash = hash;
```

Comment:
I realize this is dead code, but we shouldn’t be reversing any hashes here.

Reply:
> I realize this is dead code, but we shouldn’t be reversing any hashes here.

Agreed.  The issue is that it's read from the user in reverse (via decode_hash) and shipped to the server, where it's read forward (not reversed, via read_hash), so it causes a mis-match.  The better change is to read_reverse when deserializing at the server, rather than what's there now.

Reply:
Related to https://github.com/libbitcoin/libbitcoin-server/pull/516